### ISSUE TITLE
How to handle event handler attributes

### ISSUE THREAD

**Issue by lukewarlow** [WICG/sanitizer-api#226]
Afaik the current plan for the sanitizer API is that there would be a list of all known attributes and from that you'd subtract an underlying blocklist of unsafe attributes, event handlers are tricky because there's lots of them. If we have a list of event handlers to remove it would be useful if it was defined separately from other attributes so that the trusted types specification could piggy back off of it.

Having said that all current ones follow the `on*` pattern, should we just carve out `on*` attributes as being unsafe and block them all? I'm no perf specialist but I imagine that would be faster to process (both WebKit and Chromium already have some paths that do the on prefix check for example).

**Comment by lukewarlow** [comment]
Has their been any discussion about allowing stuff like regex in the configs? Not needed for a v1 but if we did allow some form of pattern based matching then it would solve the fact we can't currently define it in syntax.

**Comment by domenic** [comment]
I worry that this restricts the space of future HTML attribute design in a way that's hard to understand. E.g. we could not use attributes named things like `once=""`, `only=""`, `online=""`, `ongoing=""`, `onto=""`.

Maybe it would still be OK as long as we'd consider it a backward-compatible change to modify the sanitization algorithm in the future. E.g. today we'd strip all attributes starting with `on`, but in the future when we add an `only=""` attribute, we strip all attributes starting with `on` except we let `only=""` through.

**Comment by annevk** [comment]
I think we do have the expectation that what is being sanitized and how will subtly change over time due to the HTML (including SVG/MathML) language evolving. New "safe" attributes getting safelisted seems in line with that.

**Comment by otherdaniel** [comment]
I'm not fond of this, and would prefer us to be precise with the event handler attributes.

With Trusted Types, we initially had a similar logic, but kept getting bug reports about sites using their own `on*` attributes that would get erroneously blocked by TT. So we eventually re-implemented this with the actual list of actual event handlers, which solves that set of problems. I'd prefer to not repeat that experience.

Examples:
* https://g-issues.chromium.org/issues/40640134#comment4
* https://g-issues.chromium.org/issues/40640134#comment5

Also, I don't quite understand why this would be needed at all: Our default-config -- where these matter -- are allow-lists, and in an allow-lists the event handlers are very much expected to be absent. That is, no `on*` attribute -- whether a browser-defined event handler or not -- should show up in any built-in config. So why would we care? Or why would it make a perf difference?

For the config pre-processing, we do need to decide whether an attribute is known or not -- otherwise one could never override the built-in default. That indeed requires a list of known attributes, including the known event handlers. But that's required only during config normalization; and that requires every single HTML attribute ever. I doubt the event handlers make that much of a difference there.

**Comment by annevk** [comment]
I do have some lingering worry here that each time a browser introduces a new event handler attribute, the sanitization won't be the same across browsers and this could result in some issues. How to square that with sites using their own `on*` attributes is unclear though.

**Comment by mozfreddyb** [comment]
If we solved this by disallowing and listing known event handlers in the "no XSS possible" mode and disallowing all handlers starting with "on" in the default, we'd basically allow browsers to bypass the default because a user might create a Sanitizer that allows a new `onfoo` attribute that the browser supports but Sanitizer is just not aware of yet.

What if we disallow all `on*` and hardcode exceptions for those known to be widely used? Seems also a bit icky imho.

What if we just disallow all `on` and tell people to use unsafe if they are unhappy with it? Luxury of designing a new API is that we can decide not just to support harmless on* attributes... (admittedly a bit dismissive of people's use cases)

**Comment by annevk** [comment]
@cure53 how does DOMPurify deal with event handler attributes?

**Comment by cure53** [comment]
> @cure53 how does DOMPurify deal with event handler attributes?

@annevk We work with an [allow-list](https://github.com/cure53/DOMPurify/blob/main/src/attrs.ts) on which event handlers are simply not present. If folks decide they want to permit them, they can do so using the [config](https://github.com/cure53/DOMPurify/blob/main/src/purify.ts#L622), i.e. via `ADD_ATTR`.

Other than that, they receive no extra treatment, and get handled as any other attribute :slightly_smiling_face:

**Comment by lukewarlow** [comment]
> What if we just disallow all on and tell people to use unsafe if they are unhappy with it?

My personal thoughts are we should just do this. Google decided against it for trusted types (which now makes it difficult to change retrospecitvely) but I think we should just take the slight annoyance from certain websites and do it for sanitizer like should have been done for trusted types.

As you say worst case they can use unsafe if they really need their badly named attribute to work.

I think we (whatwg) should also agree to leave onXYZ to only event handlers and if someone wants some other attribute to start with 'on' we simply get them to rename it.

**Comment by otherdaniel** [comment]
As Luke said, Trusted Types initially blocked any `"on*"`, but we got too many complaints about that and then changed to an actual list, which is derived from the IDLs.

Examples problems are:
- https://g-issues.chromium.org/issues/40640134#comment9
- https://g-issues.chromium.org/issues/40640134#comment10
- https://g-issues.chromium.org/issues/40640134#comment11 
- https://g-issues.chromium.org/issues/40640134#comment12

I think Sanitizer has an easier time with this, because there's always the "Unsafe" variants. But that sure isn't pretty.

As it presently stands the defaults would be an allow list, where this isn't a problem because no event handler is going to be in the default allow list anyhow. But `removeUnsafe`, plus the implied `removeUnsafe` in safe versions, would make this observable.

**Comment by mozfreddyb** [comment]
A couple of the linked issues are not public. I _hope_ we can get away with this by a) codifying that `on` is for event handlers and assumed unsafe and b) supporting this anyway with the unsafe methods :-)

I understand that we would be dismissive of existing code and use cases for custom attributes... 😕

**Comment by annevk** [comment]
I'm a bit surprised that it seems the Trusted Types decision was made unilaterally by Chromium. Was there no discussion about this in a specification setting?

**Comment by cure53** [comment]
> I mean, the alternative is to somewhere have a canonical list of event handler attributes but that seems tricky in two ways. First, it appears that there is no such list and the HTML spec might not be the best source to find all of them. Secondly, most implementations don't really know or can figure out at build time either.

Historically, as far as I know, there was only one case where a non-event handler started with "on" and if I remember correctly, that was for the [VML implementation in MSIE](https://learn.microsoft.com/en-us/windows/win32/vml/msdn-online-vml-oned-attribute). 

Other than that, I am not aware of any attributes starting with "on" that do _not_ trigger JavaScript execution.

**Comment by mozfreddyb** [comment]
Well, see links above. Lots of frameworks do funky stuff with `on=` or `one=`, that do not lead to script execution (but may cause script gadget funkiness)

**Comment by lukewarlow** [comment]
The current behaviour also makes it pretty hard to test trusted types coverage of event handlers (and would similarly be difficult for sanitizer). You can read from spec web idl but not all browsers implement all event handlers so that can show up failures that aren't failures. It also doesn't cover any non-standard event handlers, nor does it necessarily catch all existing ones if they're not on the IDL constructs you expect.

You can maybe loop through the element classes on the global and from their loop through their properties that start with 'on' and test against each of those, by converting the class name to a tag name creating the element and trying to set the attribute. Though just because the property exists on the prototype doesn't mean it actually executes script as a content attribute.

**Comment by otherdaniel** [comment]
> The current behaviour also makes it pretty hard to test trusted types coverage of event handlers (and would similarly be difficult for sanitizer). You can read from spec web idl but not all browsers implement all event handlers so that can show up failures that aren't failures. It also doesn't cover any non-standard event handlers, nor does it necessarily catch all existing ones if they're not on the IDL constructs you expect.

I don't understand why this is hard to test. My usual test pattern is to pick a few block and a few non-block cases and test those. If we want to test exhaustively -- whatever that would even mean for the `"on*"` case -- I'd take a list of candidates and check whether they're event handlers to determine what the expected outcome under TT would be.

(Something like: ```div.setAttribute(probe, "return"); const expect_to_throw_with_tt = (typeof div[probe] == "function");``` The underlying behaviour -- converting a string to a function, or not -- is spec mandated, I believe.)

I think for IDL tests, WPT runs tests against spec-derived baseline, and whoever does something differently will have to maintain a list of expectations themselves. That puts the maintenance burden on whoever does something different. That might also work.

I wish we could somehow re-use or hook into the webref / IDL tests. That seems like a fairly comprehensive and well-tested mechanism for testing APIs against a spec baseline. I'm admittedly not seeing an obvious way to do that.

I do understand the interop concern; but don't really understand the testing one.

**Comment by noamr** [comment]
Can we rely on WebIDL somehow here? As in:
- By default, in unsafe mode, reject all the `on*` properties that are listed in WebIDLs known to the browser
- Allow some WebIDL keyword to opt-out of this behavior, e.g. `[Safe]` or whatever if we ever want to introduce attributes like `once`  
- This wouldn't reject author-defined `on*`, because those don't come from WebIDL

**Comment by annevk** [comment]
@otherdaniel FWIW, I have been paying attention since 2019 at least, from a cursory look at old issues in the relevant repository, but I did miss that the specification was already requiring that.

I've been trying to think of how to best decide on this issue. I think it comes down to how likely it is that sanitized content in one browser ends up being reflected in another. While that concern also impacts other features, event handler attributes are much more commonly introduced. Overall though that seems like a fairly minor concern and any website that does it that way will also run into other issues I suspect. So I think I'm leaning towards using the same backing list as Trusted Types.

**Comment by mozfreddyb** [comment]
We discussed this in our sanitizer meeting and found both approaches doable. However, we agreed on the following:

- in the spec, use the lists of _specified_ event handler attributes from HTML
- in implementations, make sure that there is an exhaustive mechanism to get all the _supported_ attributes on top of that
- provide no defense-in-depth for authors that sanitize in one browser and then use the result in another browser, because we can only ever match the parsing/sanitization rules of the current browser.

**Comment by mozfreddyb** [comment]
This should be solved with #244 / #243, no?
